### PR TITLE
upcoming: [M3-8287] - Add New Search Parser

### PR DESCRIPTION
**Pull request by bnussman-akamai**
## Description 📝

- Kicks off _Search v2 ✨_ by adding a "search query to API filter" parser
  - New search parser is in it's own package to prepare for micro-frontends 📦 
  - Keep in mind, this new code isn't "final" at all. We will iterate and document this over time as we improve search throughout Cloud Manager 🔍 
- Implements new parser in Linode Create v2 for StackScript filtering 📃 
  - This achieves parity with the old flow plus all of the new operations like and, or, etc..

## Preview 📷

Screenshot 2024-06-25 at 5 39 39 PM

## How to test 🧪

- Check basics of new package structure 📦 
- Check functionality of new search 🔍 
  - Turn on Linode Create v2
  - Test search on `http://localhost:3000/linodes/create?type=StackScripts`
  - Verify search functionality is better than or equivalent to Linode Create v1

## As an Author I have considered 🤔

- [ ] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [ ] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [ ] 🔐 Removing all sensitive information from the code and PR description
- [ ] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [ ] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

### CODE REVIEW

File: packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx [hunk 1]
```diff
@@ -120,8 +140,38 @@ export const StackScriptSelectionList = ({ type }: Props) => {
...
+                size="small"
+              >
+                <CloseIcon />
+              </IconButton>
+            </InputAdornment>
+          ),
+        }}
+        tooltipText={
+          type === 'Community'
+            ? 'Hint: try searching for a specific item by prepending your search term with "username:", "label:", or "description:"'
```
Comment (bnussman-akamai) [inline packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx:164]: Eventually, we'll probably have better docs/helper text for search, but for now, this is just a copy-paste from Linode Create v1 for the sake of parity

File: packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx [hunk 2]
```diff
@@ -64,17 +75,26 @@ export const StackScriptSelectionList = ({ type }: Props) => {
       ? communityStackScriptFilter
       : accountStackScriptFilter;
 
+  const {
+    error: searchParseError,
+    filter: searchFilter,
+  } = getAPIFilterFromQuery(query, {
+    defaultSearchKeys: ['username', 'label', 'description'],
+  });
```
Comment (bnussman-akamai) [inline packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx:83]: This is all it takes to get an API filter from a search query 🔥 🔍

File: packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx [hunk 3]
```diff
@@ -36,6 +45,8 @@ interface Props {
 }
 
 export const StackScriptSelectionList = ({ type }: Props) => {
+  const [query, setQuery] = useState<string>();
```
Comment (bnussman-akamai) [inline packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx:48]: We could consider using query params in the future, but my main focus is to get Linode Create v2 in parity with v1 and v1 does not use query params.

File: packages/manager/src/env.d.ts [hunk 4]
```diff
@@ -43,7 +43,7 @@ declare module '*.svg' {
   export default src;
 }
 
-declare module '*.css?raw' {
+declare module '*?raw' {
```
Comment (abailly-akamai) [inline packages/manager/src/env.d.ts:46]: What is this for?

File: packages/search/src/search.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import { generate } from 'peggy';
+import type { Filter } from '@linode/api-v4';
+import grammar from './search.peggy?raw';
+
+const parser = generate(grammar);
+
+interface Options {
+  /**
+   * Allows you to set the default keys that get search when the user types
+   * a query without an operator.
```
Comment (abailly-akamai) [inline packages/search/src/search.ts:10]:
```suggestion
   * Allows to set searchable API fields when no operator is provided in the query. 
```

File: packages/search/src/search.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+const parser = generate(grammar);
+
+interface Options {
+  /**
+   * Allows you to set the default keys that get search when the user types
+   * a query without an operator. 
+   * 
+   * @example ['label', 'tags']
+   */
+  defaultSearchKeys: string[];
```
Comment (abailly-akamai) [inline packages/search/src/search.ts:14]: I think this name is not very clear. Key is essentially matching an API field so we could be more explicit here. And "default" does not add much meaning either. ex: 

`searchableFieldsWithoutOperator: string[];`

it's a bit wordy but i like more and more explicit prop names and since  this isn't out yet it's a good time to set this right.

File: packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx (hunk 1 above)
Comment (abailly-akamai) [reply packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx:164]: This feels kinda weird because you're asking the user to do more work here. Those fields are searchable without adding `term:`. It would make more sense to tell the user they can search for any of those three fields.

File: packages/search/README.md [hunk 7]
```diff
@@ -0,0 +1,43 @@
+# Search
+
+Search is a parser written with [Peggy](https://peggyjs.org) that takes a human readable search query and transforms it into a [Linode API v4 filter](https://techdocs.akamai.com/linode-api/reference/filtering-and-sorting). 
+
+The goal of this package is to provide a shared utility that enables a powerful, scaleable, and  consistent search experience throughout Linode Cloud Manager.
```
Comment (abailly-akamai) [inline packages/search/README.md:5]:
```suggestion
The goal of this package is to provide a shared utility that enables a powerful, scalable, and  consistent search experience throughout Linode Cloud Manager.
```

File: packages/search/README.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+| `and`     | `&`, `&&`      | `label: prod and size > 20`    | Performs a boolean *and* on two expressions                     |
+| `or`      | `|`, `||`      | `label: prod or size > 20`     | Performs a boolean *or* on two expressions                      |
+| `>`       | None           | `size > 20`                    | Greater than                                                    |
+| `<`       | None           | `size < 20`                    | Less than                                                       |
+| `>=`      | None           | `size >= 20`                   | Great than or equal to                                          |
+| `<=`      | None           | `size <= 20`                   | Less than or equal to                                           |
+| `!`       | `-`            | `!label = my-linode-1`         | Not equal to (does not work as a *not* for boolean expressions) |
+| `=`       | None           | `label = my-linode-1`          | Equal to                                                        |
+| `:`       | `~`            | `label: my-linode`             | Contains                                                        |
```
Comment (abailly-akamai) [inline packages/search/README.md:43]: Any thoughts in adding this `.md` to our [Cloud Manager docs](https://linode.github.io/manager)?

File: packages/search/src/search.peggy [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+String "search value"
+  = [a-zA-Z0-9\-\.]+ { return text(); }
+  
+Number "numeric search value"
+    = number:[0-9\.]+ { return parseFloat(number.join("")); }
+    / number:[0-9]+ { return parseInt(number.join(""), 10); }
+
+ws "whitespace"
+  = [ \t\r\n]
```
Comment (abailly-akamai) [inline packages/search/src/search.peggy:99]: something new!

File: packages/search/src/search.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
+import { generate } from 'peggy';
+import type { Filter } from '@linode/api-v4';
+import grammar from './search.peggy?raw';
```
Comment (abailly-akamai) [inline packages/search/src/search.ts:3]: what's `?raw`? a comment may be useful for maintainers

File: packages/search/src/search.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  let error: SyntaxError | null = null;
+
+  try {
+    filter = parser.parse(query, options);
+  } catch (e) {
+    error = e as SyntaxError;
+  }
+
+  return { filter, error };
+}
```
Comment (abailly-akamai) [inline packages/search/src/search.ts:35]: Couple things here:
- Any reason not to use an arrow function?
- Candid (prob unwarranted scope worry): any risk of mutation if we have two searchable fields on the page?

**Review by abailly-akamai — COMMENTED** [review]
This is really cool! Great job on delivering this tool.

I have to spend more time with it, here are some initial comments for consideration

File: packages/manager/src/env.d.ts (hunk 4 above)
Comment (bnussman-akamai) [reply packages/manager/src/env.d.ts:46]: This imports a file as a string. (see [docs](https://vitejs.dev/guide/assets#importing-asset-as-string))

I incorrectly put the `css` part when I first added the type. It works on any file extension.

Usually `/// <reference types="vite/client" />` would setup this type for us, but when we migrated to Vite, I never was able to get that working.

File: packages/search/src/search.ts (hunk 10 above)
Comment (bnussman-akamai) [reply packages/search/src/search.ts:3]: `?raw` is a Vite feature that allows us to import things as strings. ([docs](https://vitejs.dev/guide/assets#importing-asset-as-string))

We do this because Peggy expects the grammar as a string

File: packages/search/src/search.ts (hunk 11 above)
Comment (bnussman-akamai) [reply packages/search/src/search.ts:35]: - I didn't use an arrow function because I don't like them. Happy to switch to an arrow function for consistency with our other repos, just let me know
- I don't think there is any concern! 🎉

Comment (abailly-akamai) [reply packages/search/src/search.ts:35]: > I didn't use an arrow function because I don't like them

I think it's ok to leave it, not a big deal, but the reason for doing it is a little too subjective 🤣

File: packages/search/README.md (hunk 8 above)
Comment (bnussman-akamai) [reply packages/search/README.md:43]: I considered that, but decided against that because, so far, we seem to use the docs site only for developer docs. This feels more relevant to customers than developers

Eventually, I'd like this to live in Cloud Manager itself or on https://www.linode.com/docs/guides/how-to-search-in-the-cloud-manager/ or both

File: packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx (hunk 1 above)
Comment (bnussman-akamai) [reply packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx:164]: This was just copy-paste from Linode Create v1 😅 

I was thinking, when we get UX involved with search v2, we can get them to refine how we let users know about the complex search capabilities

Comment (abailly-akamai) [reply packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx:164]: but.... do we have to copy the same thing over now that they don't have to add the term?? This isn't parity anymore

Comment (bnussman-akamai) [reply packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx:164]: From what I understand, on Linode Create v1, you don't have to add the term. Same applies here

Comment (abailly-akamai) [reply packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx:164]: Ok so this hint is misleading. I'll leave it at that I guess

File: packages/search/README.md [hunk 12]
```diff
@@ -0,0 +1,43 @@
+# Search
+
+Search is a parser written with [Peggy](https://peggyjs.org) that takes a human readable search query and transforms it into a [Linode API v4 filter](https://techdocs.akamai.com/linode-api/reference/filtering-and-sorting). 
+
+The goal of this package is to provide a shared utility that enables a powerful, scalable, and consistent search experience throughout Linode Cloud Manager.
```
Comment (dwiley-akamai) [inline packages/search/README.md:5]: `Linode Cloud Manager` --> `Akamai Connected Cloud Manager` to be consistent with our naming in the [README](https://github.com/linode/manager)?

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -185,6 +185,21 @@ jobs:
...
+        with:
+          node-version: "18.14"
+      - uses: actions/cache@v3
+        with:
+          path: |
+            **/node_modules
+          key: ${{ runner.os }}-${{ hashFiles('**/yarn.lock') }}
+      - run: yarn --frozen-lockfile
+      - run: yarn workspace @linode/search run test
+
```
Comment (dwiley-akamai) [inline .github/workflows/ci.yml:202]: Looks like this will run pretty quickly, which is good

Screenshot 2024-06-26 at 4 25 29 PM

**Review by dwiley-akamai — COMMENTED** [review]
Still doing some testing, but it looks good so far in terms of parity and extended functionality 🎉

File: packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx [hunk 14]
```diff
@@ -120,8 +140,38 @@ export const StackScriptSelectionList = ({ type }: Props) => {
...
+                onClick={() => setQuery('')}
+                size="small"
+              >
+                <CloseIcon />
+              </IconButton>
+            </InputAdornment>
+          ),
+        }}
+        tooltipText={
+          type === 'Community'
```
Comment (dwiley-akamai) [inline packages/manager/src/features/Linodes/LinodeCreatev2/Tabs/StackScripts/StackScriptSelectionList.tsx:163]: I know it's like this in prod, but I think the tooltip should be displayed for the Account StackScripts tab too since the operators work there as well. It makes less sense to include `username` for that tab's tooltip though.

**Review by dwiley-akamai — APPROVED** [review]
Search functionality at parity with or improved compared to prod ✅
Code review ✅

I did notice an odd behavior where errors for multiple sections pop up if you type a search term in and then press enter:

<details>
<summary>Video</summary>

https://github.com/linode/manager/assets/114682940/45e23f1c-8a12-4596-b211-9b059ea17cb1
</details>

**Review by abailly-akamai — APPROVED** [review]
Let's gooo! thanks for addressing those changes. All looks good to me code-wise and filtering works as expected ✅
